lib/Dialect/D2M/Transforms/LowerToLayout.cpp

```diff
@@ -395,19 +198,309 @@ class D2MSplitCompoundLayoutRewriter : public OpRewritePattern<ToLayoutOp> {
+    llvm::errs() << "      lowerMappingChange: output = " << output.getType()
+                 << "\n";
+
+    Value viewOp = rewriter.create<ViewLayoutOp>(loc, viewType, input,
+                                                 /*reinterpretLayout=*/false);
+
+    // Materialize L1→L1 transformations with a DMA generic that performs the
+    // actual data movement according to the view's affine map.
+    if (!inputInfo.isDRAM() && !outputInfo.isDRAM()) {
+      auto gridShape = outputInfo.getGridShape();
+
+      // If the output has a virtual grid (and by our early check, input does
+      // too with the same grid shape), we need to include the virtual→physical
+      // coordinate translation in the grid attribute.
+      ttcore::GridAttr grid;
+      if (outputHasVirtualGrid) {
+        // Check if this is actually a virtual grid (grid shape exceeds physical
+        // bounds)
+        bool isVirtualGrid = gridShape[0] > targetGridShape[0] ||
+                             gridShape[1] > targetGridShape[1];
+
+        if (isVirtualGrid) {
+          // Create the virtual grid coordinate maps and use the inverse map
+          // for the grid's coordinate translation.
+          auto [fwdMap, invMap] = ttmlir::d2m::utils::grids::createCoreVirtMaps(
+              rewriter.getContext(), gridShape, targetGridShape);
+          grid =
+              ttcore::GridAttr::get(rewriter.getContext(), gridShape, invMap);
+        } else {
+          // Grid fits within physical bounds, no coordinate translation needed
+          grid = ttcore::GridAttr::get(rewriter.getContext(), gridShape);
+        }
+      } else {
+        grid = ttcore::GridAttr::get(rewriter.getContext(), gridShape);
+      }
+
+      const size_t gridRank = gridShape.size();
+
+      // Build identity indexing maps for the generic operation. The view's
+      // affine map handles all address transformations.
+      ArrayAttr indexingMaps, iteratorTypes;
+      std::tie(indexingMaps, iteratorTypes) =
+          GenericOp::buildParallelAffineMapsAndIteratorTypes(
+              rewriter, /*arity=*/2, gridRank);
+      auto indexingMap = mlir::cast<AffineMapAttr>(indexingMaps[0]);
+
+      return rewriter
+          .create<GenericOp>(
+              loc, viewOp, output,
+              [&](OpBuilder &builder, Location innerLoc, ValueRange blockArgs) {
+                Value outputCB =
+                    builder.create<ReserveOp>(innerLoc, blockArgs[1])
+                        .getResult();
+                auto dma = builder.create<d2m::DMAOp>(innerLoc, viewOp,
+                                                      indexingMap, outputCB);
+                builder.create<d2m::DMAWaitOp>(innerLoc, dma);
+                builder.create<YieldOp>(innerLoc, outputCB);
+              },
+              ThreadType::Datamovement, grid)
+          .getResult(0);
+    } else {
+      // DRAM operations use the view directly without immediate
```

Comment:
:warning: **llvm\-else\-after\-return** :warning:
do not use `` else `` after `` return ``
```suggestion
    }       // DRAM operations use the view directly without immediate
```